import dedent from 'dedent-js'
import Layout from '../components/Layout'
import Notice from '../components/Notice'
import TabbedCodeExamples from '../components/TabbedCodeExamples'

export default Layout
export const meta = {
  title: 'Error handling',
  links: [{ url: '#in-development', name: 'In development' }, { url: '#in-production', name: 'In production' }],
}

# Error handling

## In development

One of the nice things about working with a server-side framework is the built-in exception handling you get for free. For example, Laravel ships with [Ignition](https://github.com/facade/ignition), a beautiful error reporting tool which displays a nicely formatted stack trace in local development.

The challenge is, if you're making an XHR request (which Inertia does), and you hit a server-side error, you're typically left digging through the network tab in your browser's devtools.

Inertia solves this by showing all non-Inertia responses in a modal. Meaning you get the same beautiful error-reporting, even though you've made that request over XHR!

<div className="my-6 relative rounded overflow-hidden bg-gray-500" style={{ paddingTop: '80.5%' }}>
  <div className="absolute inset-0 w-full h-full flex items-center justify-center text-sm">Loading&hellip;</div>
  <iframe
    className="absolute inset-0 w-full h-full"
    src="https://player.vimeo.com/video/363562630?autoplay=1&loop=1&muted=1&background=1"
  ></iframe>
</div>

<Notice>Note, the modal behaviour is only intended for development purposes.</Notice>

## In production

In production you'll want to return a proper Inertia error response instead of relying on the modal behaviour. To do this you'll need to update your framework's default exception handler to return a custom error page.

<TabbedCodeExamples
  examples={[
    {
      name: 'Laravel',
      language: 'php',
      code: dedent`
        use Inertia\\Inertia;\n
        public function render($request, Exception $exception)
        {
            $response = parent::render($request, $exception);\n
            if ($request->header('X-Inertia') && in_array($response->status(), [500, 503, 404, 403])) {
                return Inertia::render('Error', ['status' => $response->status()])
                    ->toResponse($request)
                    ->setStatusCode($response->status());
            }\n
            return $response;
        }
      `,
    },
    {
      name: 'Rails',
      language: 'ruby',
      code: dedent`
        # app/controllers/exceptions_controller.rb
        class ExceptionsController < ActionController::Base
          def show
            render inertia: 'Error', props: {
              status: request.path_info[1..-1].to_i
            }
          end
        end
        
        # config/application.rb
        module MyApplication
          class Application < Rails::Application
            # ...
            config.exceptions_app = ->(env) { ExceptionsController.action(:show).call(env) }
          end
        end
        
        # Delete static error pages
        # $ rm public/404.html
        # $ rm public/422.html
        # $ rm public/500.html
      `,
    },
  ]}
/>

Notice how we're returning an `Error` page component in the example above. You'll need to actually create this. Here's an example error page component you can use as a starting point.

<TabbedCodeExamples
  examples={[
    {
      name: 'Vue.js',
      language: 'twig',
      code: dedent`
        <template>
          <div>
            <h1>{{ title }}</h1>
            <div>{{ description }}</div>
          </div>
        </template>\n
        <script>
        export default {
          props: {
            status: Number,
          },
          computed: {
            title() {
              return {
                503: '503: Service Unavailable',
                500: '500: Server Error',
                404: '404: Page Not Found',
                403: '403: Forbidden',
              }[this.status]
            },
            description() {
              return {
                503: 'Sorry, we are doing some maintenance. Please check back soon.',
                500: 'Whoops, something went wrong on our servers.',
                404: 'Sorry, the page you are looking for could not be found.',
                403: 'Sorry, you are forbidden from accessing this page.',
              }[this.status]
            },
          },
        }
        </script>
      `,
    },
    {
      name: 'React',
      language: 'jsx',
      code: dedent`
        import React from 'react'\n
        export default function ErrorPage({ status }) {
          const title = {
            503: '503: Service Unavailable',
            500: '500: Server Error',
            404: '404: Page Not Found',
            403: '403: Forbidden',
          }[status]\n
          const description = {
            503: 'Sorry, we are doing some maintenance. Please check back soon.',
            500: 'Whoops, something went wrong on our servers.',
            404: 'Sorry, the page you are looking for could not be found.',
            403: 'Sorry, you are forbidden from accessing this page.',
          }[status]\n
          return (
            <div>
              <h1>{title}</h1>
              <div>{description}</div>
            </div>
          )
        }
      `,
    },
    {
      name: 'Svelte',
      language: 'html',
      code: dedent`
        <script>
          export let status\n
          $: title = {
            503: '503: Service Unavailable',
            500: '500: Server Error',
            404: '404: Page Not Found',
            403: '403: Forbidden',
          }[status]\n
          $: description = {
            503: 'Sorry, we are doing some maintenance. Please check back soon.',
            500: 'Whoops, something went wrong on our servers.',
            404: 'Sorry, the page you are looking for could not be found.',
            403: 'Sorry, you are forbidden from accessing this page.',
          }[status]
        </script>\n
        <div>
          <h1>{title}</h1>
          <div>{description}</div>
        </div>
      `,
    },
  ]}
/>
